pinot-query-planner/src/main/java/org/apache/pinot/query/parser/utils/ParserUtils.java

```diff
@@ -53,95 +54,87 @@ public static boolean canCompileWithMultiStageEngine(String query, String databa
+        ColumnDataType columnDataType;
+        try {
+          columnDataType = RelToPlanNodeConverter.convertToColumnDataType(dataTypeField.getType());
+        } catch (Exception ignored) {
+          columnDataType = ColumnDataType.UNKNOWN;
         }
-        responseSchema.getColumnDataTypes()[i] = resolved;
+        columnDataTypes[i] = columnDataType;
       }
-    } else if (hasV1Schema) {
-      // match v1 column types with v1 schema columns using column names
-      // if no match, just leave it as it is
-      DataSchema responseSchema = response.getResultTable().getDataSchema();
-      for (int i = 0; i < responseSchema.size(); i++) {
-        FieldSpec spec = schema.getFieldSpecFor(responseSchema.getColumnName(i));
+      response.setResultTable(new ResultTable(new DataSchema(columnNames, columnDataTypes), List.of()));
+      return;
+    }
+
+    // When data schema is available, try to fix the data types within it.
+    ColumnDataType[] columnDataTypes = dataSchema.getColumnDataTypes();
+    int numColumns = columnDataTypes.length;
+    if (dataTypeFields != null) {
+      // Fill data type with the validated row type when it is available.
+      for (int i = 0; i < numColumns; i++) {
         try {
-          // try single value first
-          resolved = DataSchema.ColumnDataType.fromDataType(spec.getDataType(), true);
-        } catch (Exception e) {
-          try {
-            // fallback to multi value
-            resolved = DataSchema.ColumnDataType.fromDataType(spec.getDataType(), false);
-          } catch (Exception e2) {
-            resolved = DataSchema.ColumnDataType.UNKNOWN;
-          }
-        }
-        if (resolved == null || resolved.isUnknown()) {
-          resolved = responseSchema.getColumnDataType(i);
+          columnDataTypes[i] = RelToPlanNodeConverter.convertToColumnDataType(dataTypeFields.get(i).getType());
+        } catch (Exception ignored) {
+          // Ignore exception and keep the type from response
         }
-        responseSchema.getColumnDataTypes()[i] = resolved;
       }
-    } else if (hasV2Schema) {
-      // trust v2 column types blindly
-      // if a type cannot be resolved, leave it as UNKNOWN
-      List<RelDataTypeField> fields = node.validatedRowType.getFieldList();
-      String[] columnNames = new String[fields.size()];
-      DataSchema.ColumnDataType[] columnDataTypes = new DataSchema.ColumnDataType[fields.size()];
-      for (int i = 0; i < fields.size(); i++) {
-        columnNames[i] = fields.get(i).getName();
-        resolved = RelToPlanNodeConverter.convertToColumnDataType(fields.get(i).getType());
-        if (resolved == null) {
-          resolved = DataSchema.ColumnDataType.UNKNOWN;
+    } else {
+      // Fill data type with the schema when validated row type is not available.
+      String[] columnNames = dataSchema.getColumnNames();
+      for (int i = 0; i < numColumns; i++) {
+        FieldSpec fieldSpec = schema.getFieldSpecFor(columnNames[i]);
```

Comment:
It was similar in the previous code, but... this can get pretty crazy if the query includes aliases, right? Something like having a table with `strCol` and `intCol` of String and int type respectively and a bizarre query like:

```
select intCol as strCol, strCol as intCol
from Table
where <impossible condition that doesn't parse in MSE>
```

Will return that `strCol` is `String` and `intCol` is `int`. I guess there is not much we can do when a SSE query doesn't parse in MSE.

Reply:
Yeah, we are basically doing best effort to fill these types. There is no guarantee all SSE queries are SQL compatible